flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/JobMaster.java

```diff
@@ -944,7 +943,13 @@ public void heartbeatFromResourceManager(final ResourceID resourceID) {
 		}
 		return checkpointCoordinator
 			.triggerSavepoint(System.currentTimeMillis(), targetDirectory)
-			.thenApply(CompletedCheckpoint::getExternalPointer)
+			.thenApply(checkpointExecuteResult -> {
+				try {
+					return checkpointExecuteResult.getCompletedCheckpoint().getExternalPointer();
+				} catch (Exception e) {
```

Comment:
This `try-catch` feels bad here, just to catch the illegal state exception. First improvement could be to check `isSuccess` instead of using exceptions for controlflow. Second, as commented before, if we just use `CompletableFuture<CompletedCheckpoint>` with the capability of `#completeExceptionally` over `CheckpointExecutionResult` we can just skip the `thenApply` and handle all cases in `handleAsync`.